### PR TITLE
C++ API and some core C++ fixes and updates.

### PR DESCRIPTION
**Pull request by RubyNova**
This commit contains both the LLAPI and HLAPI, as well as some core changes being made to the rendering core as needs change.

Tasks (Core):
- [x] Fix layer **rendering** system
- [x] Fix layer **update** system
- [x] Implement input system
- [ ] Add missing tools such as textboxes and buttons 

Tasks (LLAPI):
- [ ] Provide all relevant game events
- [ ] Provide input system hooks
- [ ] Provide all object creation methods as they are created in the core (buttons, textboxes, etc.)
- [ ] Ensure extern C wrapper friendliness

Tasks (HLAPI):
- [ ] An equivalent of Unity's `MonoBehaviour` is required, only not so expensive to use. It will be based on the events defined in the LLAPI.

### CODE REVIEW

**Comment by Erlite** [comment]
https://github.com/novelrt/NovelRT/blob/d35ddbdf958f3ac93bcd73fe5b4e150e8de70db9/src/NovelRenderingService.h#L38

Typo on the function.

**Comment by RubyNova** [comment]
Work is being merged early due to a change of plans.

File: src/GeoBounds.cpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
+//
+// Created by matth on 30/03/2019.
+//
+
+#include "GeoBounds.h"
+namespace NovelRT {
+GeoBounds::GeoBounds(const NovelRT::GeoVector<float>& position, const NovelRT::GeoVector<float>& size) {
+  _corners[0] = GeoVector<float>(position.getX() - (size.getX() / 2), position.getY() - (size.getY() / 2));
```
Comment (Tylertron1998) [inline src/GeoBounds.cpp]: could this not be a for loop?

File: src/GeoBounds.cpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+namespace NovelRT {
+GeoBounds::GeoBounds(const NovelRT::GeoVector<float>& position, const NovelRT::GeoVector<float>& size) {
+  _corners[0] = GeoVector<float>(position.getX() - (size.getX() / 2), position.getY() - (size.getY() / 2));
+  _corners[1] = GeoVector<float>(position.getX() + (size.getX() / 2), position.getY() - (size.getY() / 2));
+  _corners[2] = GeoVector<float>(position.getX() + (size.getX() / 2), position.getY() + (size.getY() / 2));
+  _corners[3] = GeoVector<float>(position.getX() - (size.getX() / 2), position.getY() + (size.getY() / 2));
+}
+
+bool GeoBounds::pointIsWithinBounds(const GeoVector<float>& point) const {
+  if (point.getX() >= _corners[0].getX() && point.getX() <= _corners[2].getX() && point.getY() <= _corners[2].getY()
```
Comment (Tylertron1998) [inline src/GeoBounds.cpp]: suggestion:
```cpp
return point.getX() >= _corners[0].getX() &&
           point.getX() <= _corners[2].getX() &&
           point.getY() <= _corners[2].getY() &&
           point.getY() >= _corners[0].getY();
```

File: src/GeoBounds.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+#ifndef NOVELRT_GEOBOUNDS_H
+#define NOVELRT_GEOBOUNDS_H
+
+#include "GeoVector.h"
+namespace NovelRT {
+class GeoBounds {
+public:
+  GeoBounds(const GeoVector<float>& position, const GeoVector<float>& size);
+  //GeoVector<float>> getCorners() const;
```
Comment (Tylertron1998) [inline src/GeoBounds.h]: is this still needed?

File: src/NovelImageRect.cpp [hunk 4]
```diff
@@ -2,42 +2,54 @@
...
+  _imageHandle = nvgCreateImage(_drawContext, _imageDir.c_str(), NVG_IMAGE_NEAREST);
+  _width = 0;
+  _height = 0;
+  nvgImageSize(_drawContext, _imageHandle, &_width, &_height);
+  setSize(GeoVector<float>(_width, _height));
+  setScale(args.startingScale);
+}
+
+void NovelImageRect::setScale(const GeoVector<float>& value) {
+  _paint = nvgImagePattern(_drawContext,
```
Comment (Tylertron1998) [inline src/NovelImageRect.cpp]: this style seems pretty inconsistent with other long parameter lines - not sure if this is how it is supposed to be, however

File: src/GeoBounds.cpp (hunk 1 above)
Comment (RubyNova) [reply src/GeoBounds.cpp]: No - the maths is different for both sides of the bounds.

File: src/NovelImageRect.cpp (hunk 4 above)
Comment (RubyNova) [reply src/NovelImageRect.cpp]: Yeah not worth fixing - nanovg will be gone come my next branch.

**Pull request merged**